### ISSUE TITLE
[Phase 4] Benchmark ZO2 vs LoRA/QLoRA on Downstream Tasks

### ISSUE THREAD

**Issue by ccchow** [ccchow/zo2#6]
## 🎯 Objective
Comprehensive comparison of ZO2 against LoRA/QLoRA on standard NLP benchmarks to validate full fine-tuning advantages.

## 📊 Comparison Framework
Compare three approaches:
1. **ZO2**: Full parameter fine-tuning with offloading
2. **LoRA**: Low-rank adaptation (r=8, 16, 32)
3. **QLoRA**: 4-bit quantized LoRA

### GLUE Tasks
- [ ] SST-2 (sentiment analysis)
- [ ] MRPC (paraphrase detection)
- [ ] CoLA (linguistic acceptability)
- [ ] RTE (textual entailment)

### Generation Tasks
- [ ] SQuAD v2 (question answering)
- [ ] CNN/DailyMail (summarization)

### Custom Tasks
- [ ] Domain adaptation benchmark
- [ ] Few-shot learning scenarios

## 🎯 Metrics to Compare
1. **Accuracy/F1 Score**
2. **Memory Usage**
3. **Training Time**
4. **Inference Speed**
5. **Parameter Efficiency**

## 🧪 Experimental Setup
```python
# Standardized configuration
base_config = {
    "model": "facebook/opt-1.3b",
    "max_epochs": 3,
    "batch_size": 8,
    "learning_rate": 1e-5,
    "eval_steps": 100
}

# ZO2 config
zo2_config = ZOConfig(
    method="mezo-sgd",
    zo2=True,
    eps=1e-3
)

# LoRA config
lora_config = LoRAConfig(
    r=16,
    lora_alpha=32,
    lora_dropout=0.1
)
```

## 📊 Expected Results Table
| Method | SST-2 Acc | Memory | Time | Params |
|--------|-----------|--------|------|--------|
| Full FT | 94.5% | 20GB | 100% | 100% |
| ZO2 | 93.0% | 3.2GB | 200% | 100% |
| LoRA-16 | 92.5% | 4GB | 50% | 0.5% |
| QLoRA | 91.0% | 2GB | 60% | 0.5% |

## 📈 Analysis Deliverables
1. Performance comparison table
2. Memory-accuracy trade-off curves
3. Training efficiency analysis
4. Best practices guide
5. Method selection flowchart

## 🔬 Hypothesis to Test
- ZO2 achieves better accuracy than LoRA on complex tasks
- ZO2 has better domain adaptation due to full fine-tuning
- Memory savings justify 2x training time

## 🔗 References
- LoRA paper: https://arxiv.org/abs/2106.09685
- QLoRA paper: https://arxiv.org/abs/2305.14314
- GLUE benchmark: https://gluebenchmark.com/

**Priority**: High
**Timeline**: 5-7 days
**Dependencies**: None (can run in parallel)
